lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -29,4 +132,273 @@
 # to 0 if you are raising the major API version
 LIBPATCH = 1
 
-# TODO: add your code here! Happy coding!
+from ipaddress import IPv4Address, IPv4Network
+from copy import deepcopy
+from typing import Dict, List, Union, TypeAlias
+from ops.framework import Object, EventSource, EventBase, ObjectEvents
+from ops.charm import CharmBase
+from ops import RelationJoinedEvent, RelationChangedEvent, RelationDepartedEvent, StoredState
+import logging, json
+
+logger = logging.getLogger(__name__)
+
+Network: TypeAlias = Dict[
+    str,  # 'network' | 'gateway' | 'routes'
+    Union[
+        IPv4Address,  # gateway, ex: '192.168.250.1'
+        IPv4Network,  # network, ex: '192.168.250.0/24'
+        List[  # List of routes
+            Dict[
+                str,  # 'destination' | 'gateway'
+                Union[
+                    IPv4Address,  # gateway, ex: '192.168.250.3'
+                    IPv4Network,  # destination, ex: '172.250.0.0/16'
+                ],
+            ]
+        ],
+    ],
+]
+
+RoutingTable: TypeAlias = Dict[
+    str,  # Name of the application
+    List[Network],  # All networks for this application
+]
+
+
+class RoutingTableUpdatedEvent(EventBase):
+    """
+    Charm event for when a host registers a route to an existing interface in the router
+    """
+
+
+class RouterProviderCharmEvents(ObjectEvents):
+    routing_table_updated = EventSource(RoutingTableUpdatedEvent)
+
+
+class RouterProvides(Object):
+    """This class is used to manage the routing table of the router provider.
+
+    It's capabilities are to:
+    * Manage the routing table in the charm itself, by adding and removing
+    new network and route requests by integrated units,
+    * Synchronize the databags of all requiring units with the router table of the
+    provider charm
+
+    Attributes:
+        charm:
+            The Charm object that instantiates this class
+        _stored:
+            The persistent state that keeps the internal routing table.
+    """
+
+    on = RouterProviderCharmEvents()
+    _stored = StoredState()
```

Comment:
Find a way not to require StoredState. This could be done by looking at all the relations and writing the routing_table. We typically want to avoid using it in all charms unless absolutely necessary (i.e. there is not other way to fetch this info).

Reply:
Done.